app/plugin.go

```diff
@@ -594,17 +595,14 @@ func (a *App) getPluginsFromFolder() (map[string]*pluginSignaturePath, *model.Ap
 		return nil, model.NewAppError("getPluginsFromDir", "app.plugin.sync.list_filestore.app_error", nil, appErr.Error(), http.StatusInternalServerError)
 	}
 	pluginSignaturePathMap := make(map[string]*pluginSignaturePath)
-	for _, path := range fileStorePaths {
-		if strings.HasSuffix(path, ".tar.gz") {
-			id := strings.TrimSuffix(filepath.Base(path), ".tar.gz")
-			helper := &pluginSignaturePath{
-				pluginId:      id,
-				path:          path,
-				signaturePath: "",
-			}
-			pluginSignaturePathMap[id] = helper
-		}
+	populatePluginSignatureMap(pluginSignaturePathMap, fileStorePaths, ".tar.gz")
+
+	cannaryTag := *a.Config().PluginSettings.CanaryTag
+	if cannaryTag != "" {
+		canarySuffix := fmt.Sprintf(".tar.gz%s", cannaryTag)
+		populatePluginSignatureMap(pluginSignaturePathMap, fileStorePaths, canarySuffix)
```

Comment:
Should we invert the check order and look for canary plugins first and then only take the remaining `.tar.gz` if they are not already installed? It seems that in this flow it will always pickup any non-canary plugins regardless if they are installed or not. 

According to the doc spec 

> If a canary tag is defined, it first considers files ending with .tar.gz+<canary_tag>. It then considers files without the canary tag only if a bundle matching the same plugin id was not already installed.

@lieut-data I'm I interpreting this correct?

Reply:
@marianunez, yes, the behaviour you describe is desired.

Reply:
We are doing exactly that. First map is populated with no canary plugins and after if there are plugins with canary tags values in map are overwritten. This way plugins with canary tags have "higher priority".

Reply:
I guess the part that is not clear for me right now is the last statement of that specification: 

> ... then considers files without the canary tag **only if a bundle matching the same plugin id was not already installed.**

Reply:
There are plugins in the store - some of them might have a canary tag, others do not. For each plugin id we are trying to install file with canary tag first if there is one. If we already installed a plugin with canary tag that's it. If not we will be installing a plugin without canary tag.